drivers/bluetooth/hci/hci_ambiq.c

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright (c) 2023 Ambiq Micro Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Ambiq SPI based Bluetooth HCI driver.
+ */
+
+#define DT_DRV_COMPAT ambiq_bt_hci_spi
+
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/bluetooth/hci_driver.h>
+#include <zephyr/bluetooth/hci.h>
+
+#define LOG_LEVEL CONFIG_BT_HCI_DRIVER_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_hci_driver);
+
+#define HCI_SPI_NODE DT_COMPAT_GET_ANY_STATUS_OKAY(ambiq_bt_hci_spi)
+#define SPI_DEV_NODE DT_BUS(HCI_SPI_NODE)
+
+#define HCI_CMD 0x01
+#define HCI_ACL 0x02
+#define HCI_SCO 0x03
+#define HCI_EVT 0x04
+
+/* Offset of special item */
+#define PACKET_TYPE         0
+#define PACKET_TYPE_SIZE    1
+#define EVT_HEADER_TYPE     0
+#define EVT_HEADER_EVENT    1
+#define EVT_HEADER_SIZE     2
+#define EVT_VENDOR_CODE_LSB 3
+#define EVT_VENDOR_CODE_MSB 4
+#define CMD_OGF             1
+#define CMD_OCF             2
+
+#define EVT_OK      0
+#define EVT_DISCARD 1
+#define EVT_NOP     2
+
+/* Max SPI buffer length for transceive operations.
+ * The maximum TX packet number is 512 bytes data + 12 bytes header.
+ * The maximum RX packet number is 255 bytes data + 3 header.
+ */
+#define SPI_MAX_TX_MSG_LEN 524
+#define SPI_MAX_RX_MSG_LEN 258
+
+static uint8_t g_hciRxMsg[SPI_MAX_RX_MSG_LEN];
+static const struct device *spi_dev = DEVICE_DT_GET(SPI_DEV_NODE);
+static struct spi_config spi_cfg = {
+	.operation = SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | SPI_MODE_CPOL | SPI_MODE_CPHA |
+		     SPI_WORD_SET(8),
+};
+static K_KERNEL_STACK_DEFINE(spi_rx_stack, CONFIG_BT_DRV_RX_STACK_SIZE);
+static struct k_thread spi_rx_thread_data;
+
+static struct spi_buf spi_tx_buf;
+static struct spi_buf spi_rx_buf;
+static const struct spi_buf_set spi_tx = {.buffers = &spi_tx_buf, .count = 1};
+static const struct spi_buf_set spi_rx = {.buffers = &spi_rx_buf, .count = 1};
+
+static K_SEM_DEFINE(sem_irq, 0, 1);
+static K_SEM_DEFINE(sem_spi_available, 1, 1);
+
+typedef int (*bt_spi_transceive_fun)(void *tx, uint32_t tx_len, void *rx, uint32_t rx_len);
+typedef int (*spi_transmit_fun)(uint8_t *data, uint16_t len);
+
+extern int bt_apollo_dev_init(void);
+extern int bt_apollo_spi_send(uint8_t *data, uint16_t len, bt_spi_transceive_fun transceive);
+extern int bt_apollo_spi_rcv(uint8_t *data, uint16_t *len, bt_spi_transceive_fun transceive);
+extern int bt_apollo_controller_init(spi_transmit_fun transmit);
+extern int bt_apollo_vnd_setup(void);
+extern bool bt_apollo_vnd_rcv_ongoing(uint8_t *data, uint16_t len);
```

Comment:
Wouldn't it be cleaner to have a proper header file for these?

Reply:
I added a header file in hal_ambiq to declare these functions to make the HCI driver cleaner.

Reply:
> I added a header file in hal_ambiq to declare these functions to make the HCI driver cleaner.

Aren't all of these APIs exposed implemented by `apollox_blue.c`? The right place for the header file would therefore be here in the same directory, not in the HAL tree.

Reply:
> Aren't all of these APIs implemented by `apollox_blue.c`? The right place for the header file would therefore be here in the same directory, not in the HAL tree.

I put the header file `apollox_blue.h` in the same directory with `apollox_blue.c`, thank you.